llm/src/hf_causal_lm.py

```diff
@@ -1,85 +1,187 @@
+        "transformer.h",
+        "model.decoder.layers",
+        "gpt_neox.layers",
+    )
+    return findattr(model, hidden_layers_attrs)
+
+
+def hf_get_tied_embedding_weights(model: torch.nn.Module) -> torch.nn.Module:
+    """Returns the embeddings, which are weight tied layers.
+
+    NOTE: Different model configurations have different embedding attribute names.
+        - wte: (GPT2LMHeadModel, GPTJForCausalLM, GPTNeoForCausalLM)
+        - word_embeddings: (BloomForCausalLM)
+        - embed_tokens: (OPTForCausalLM)
+        - GPT NeoX doesn't weight tie
+    """
+    tied_embedding_attrs = (
+        "wte",
+        "word_embeddings",
+        "embed_tokens",
+    )
+    return findattr(model, tied_embedding_attrs)
+
+# /end helper functions
+
+
+def prepare_hf_causal_lm_model_for_fsdp(model: AutoModelForCausalLM):
+    """FSDP wrap a HuggingFace model
+
+    Wrap any model for FSDP which follows one of the 3 existing conventions from
+    HuggingFace for decoder-only LLMs.
+    """
+    causal_base_model = hf_get_causal_base_model(model)
+    model_block = hf_get_causal_hidden_layers(model)[0]
+    block_type = type(model_block)
+    lm_head = hf_get_causal_hidden_layers(model)
+    tied_embeddings = hf_get_tied_embedding_weights(causal_base_model)
+    modules = [causal_base_model, model_block, block_type, lm_head, tied_embeddings]
+    if not all(module is not None for module in modules):
+        raise ValueError('Unable to FSDP-wrap this model! It does not follow \
+                          common layer/weight naming conventions.')
+    # When using the HF LM models,
     # the weights of the self.lm_head and self.transformer.wte are tied.
     # This tying occurs inside the `self.post_init()` function.
     # This is a hurdle for FSDP because they need to be in the same FSDP block
-    # These lines ensures that both modules stay together in the top-most block
-    model.transformer._fsdp_wrap = False
-    model.transformer.wte._fsdp_wrap = False
-    model.lm_head._fsdp_wrap = False
+    # These lines ensures that both modules stay together in the top-most block when
+    # the model has this tying enabled (almost all do; this property defaults to True)
+    if model.config.tie_word_embeddings:
+        causal_base_model._fsdp_wrap = False
+        tied_embeddings._fsdp_wrap = False
+        lm_head._fsdp_wrap = False
 
-    # FSDP Wrap and Activation Checkpoint every GPT2Block
-    model.fsdp_wrap_fn = lambda module: isinstance(module, _HF_MODEL_BLOCKS)
+    # FSDP Wrap and Activation Checkpoint every model block
+    model.fsdp_wrap_fn = lambda module: isinstance(module, block_type)
     model.activation_checkpointing_fn = lambda module: isinstance(
-        module, _HF_MODEL_BLOCKS)
+        module, block_type)
 
 
-class ComposerHFCausalLM(ComposerModel):
-
-    def __init__(self, cfg):
-        super().__init__()
+class ComposerHFCausalLM(HuggingFaceModel):
+    def __init__(self, cfg: DictConfig):
```

Comment:
Looking at the flow here, I think we are taking a `cfg`, creating some HF objects, then initializing a `HuggingFaceModel`.  Could this be turned into a function rather than a subclass? Basically, this would demonstrate that `HuggingFaceModel` is usable without requiring writing your own ComposerModel class.

Also, if this is not possible, then we will have some feedback for Composer team :)

Reply:
So I tried that, and that is how I ended up with those mis-shapen tensor errors. I had to override `loss`, `update_metric`, like was being done before in the class that wrapped `ComposerModel`.

Reply:
I think once I merge https://github.com/mosaicml/composer/pull/1857 we might be able to do this. (I think) RN the reshaping is necessary because you are using `Perplexity` which uses `HFCrossEntropy` which doesn't reshape for you.

Reply:
there might still remain a difference in the loss bit (our `HuggingFaceModel` assumes that the underlying HF model returns the loss as part of its output), but i'd have to look into our implementation to check that.

Reply:
Thanks @dakinggg ! I think in that case, we can merge as is for the functionality, with a TODO to cleanup in the future, so that we can use `HuggingFaceModel` directly.